### PR TITLE
[CodeCompletion] Avoid re-typechecking pre-checked expressions

### PR DESCRIPTION
**Pull request by rintaro**
Implicit call expressions for "KeyPath expression as functions" weren't sanitized in `SanitizeExpr`. That leaves `AutoClosureExpr` in the AST before re-typechecking, and caused crashes.

Handle them in `SanitizeExpr`. (the first commit)

But we don't need to re-typecheck them in the first place. (the second commit)

rdar://problem/60982638

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@xedin Could you review `SanitizeExpr` part? This is now unreachable as far as we know, but it should avoid similar crashes in future.

**Review by xedin — APPROVED** [review]
LGTM!

**Comment by slavapestov** [comment]
If SanitizeExpr is unreachable, we should just remove it. There's no way to make it correct for the general case, and it has itself caused bugs in the past.

**Comment by xedin** [comment]
One we integrate code completion into solutions with fixes scheme we can finally deprecate `SanitizeExpr` since there is going to be no re-typechecking left.

**Comment by rintaro** [comment]
@slavapestov Unfortunately we cannot remove `SanitizeExpr` yet because we still have some code path that calls re-typechecking. If we are 100% sure that normal compilations never do re-typechecking, we can move `SanitizeExpr` out of CSGen and call it manually.

File: lib/IDE/ExprContextAnalysis.cpp [hunk 1]
```diff
@@ -393,12 +393,16 @@ static void collectPossibleCalleesByQualifiedLookup(
...
   }
 
-  auto baseTyOpt = getTypeOfCompletionContextExpr(
-      DC.getASTContext(), &DC, CompletionTypeCheckKind::Normal, baseExpr, ref);
-  if (!baseTyOpt)
-    return;
-
-  auto baseTy = (*baseTyOpt)->getWithoutSpecifierType();
+  Type baseTy = baseExpr->getType();
+  if (!baseTy || baseTy->is<ErrorType>()) {
```
Comment (benlangmuir) [inline lib/IDE/ExprContextAnalysis.cpp:397]: Why is it safe to re-check when it's ErrorType?

Comment (rintaro) [reply lib/IDE/ExprContextAnalysis.cpp:397]: It *should* be safe for any expression as long as `SanitizeExpr` works correctly.
This change is kind of an optimization which avoid unnecessary re-typechecking.

Comment (benlangmuir) [reply lib/IDE/ExprContextAnalysis.cpp:397]: So this change does not affect the correctness of this patch?

Comment (rintaro) [reply lib/IDE/ExprContextAnalysis.cpp:397]: Right. The first patch fixes the crash. The second patch doesn't affect the result of it.

**Review by benlangmuir — APPROVED** [review]

**Pull request merged**